File: scripts/mirror-prod-to-sandbox.mjs
#!/usr/bin/env node
/**
 * Mirror production data to the sandbox Supabase project.
 *
 * Runs through Supabase's management API (no DB password needed).
 *
 * Strategy per table:
 *   1. On sandbox, query information_schema to get the list of
 *      non-generated, non-identity-always columns we're allowed to
 *      insert into.
 *   2. On prod, SELECT jsonb_agg of those specific columns.
 *   3. On sandbox, TRUNCATE-like cleanup + INSERT from the JSON
 *      restricted to the column list.
 *
 * `SET session_replication_role = replica` disables triggers + FK
 * checks during the load so order doesn't matter and triggers don't
 * fire twice for the same row.
 *
 * Sandbox-only tables (modules_catalog, goals_*, etc.) are left
 * untouched — they don't exist on prod and don't appear in TABLES.
 */

const TOKEN = process.env.SUPABASE_ACCESS_TOKEN;
const PROD = process.env.PROD_REF || "bwtmtpcgilvrkhougjdo";
const SANDBOX = process.env.SANDBOX_REF || "hqsdhwlokfwcitfitees";

if (!TOKEN) {
  console.error("SUPABASE_ACCESS_TOKEN env var required");
  process.exit(1);
}

const TABLES = [
  { schema: "auth", name: "users" },
  { schema: "auth", name: "identities" },
  { schema: "public", name: "profiles" },
  { schema: "public", name: "spaces" },
  { schema: "public", name: "space_members" },
  { schema: "public", name: "terminals" },
  { schema: "public", name: "terminal_members" },
  { schema: "public", name: "tasks" },
  { schema: "public", name: "task_assignees" },
  { schema: "public", name: "task_dependencies" },
  { schema: "public", name: "files" },
  { schema: "public", name: "comments" },
  { schema: "public", name: "message_threads" },
  { schema: "public", name: "thread_participants" },
  { schema: "public", name: "messages" },
  { schema: "public", name: "activity" },
  { schema: "public", name: "calendar_connections" },
  { schema: "public", name: "calendar_events" },
  { schema: "public", name: "invites" },
  { schema: "public", name: "notifications" },
  { schema: "public", name: "access_tokens" },
  { schema: "public", name: "api_keys" },
];

async function query(ref, sql) {
  const r = await fetch(
    `https://api.supabase.com/v1/projects/${ref}/database/query`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${TOKEN}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ query: sql }),
    },
  );
  if (!r.ok) {
    const body = await r.text();
    throw new Error(`HTTP ${r.status} ${body.slice(0, 400)}`);
  }
  return r.json();
}

function sqlIdent(s) {
  return '"' + s.replaceAll('"', '""') + '"';
}
function sqlString(s) {
  return "'" + s.replaceAll("'", "''") + "'";
}

async function tableExists(ref, schema, name) {
  const r = await query(
    ref,
    `SELECT to_regclass('${schema}.${name}') IS NOT NULL AS exists`,
  );
  return r?.[0]?.exists === true;
}

/**
 * Insertable columns on a table — excludes generated columns and
 * identity-always columns. Returns the list in column ORDER (matters
 * for jsonb_populate_record path).
 */
async function insertableColumns(ref, schema, name) {
  const r = await query(
    ref,
    `SELECT column_name, ordinal_position
     FROM information_schema.columns
     WHERE table_schema = ${sqlString(schema)}
       AND table_name = ${sqlString(name)}
       AND is_generated = 'NEVER'
       AND identity_generation IS DISTINCT FROM 'ALWAYS'
     ORDER BY ordinal_position`,
  );
  return r.map((row) => row.column_name);
}

async function dumpTable(schema, name, cols) {
  const exists = await tableExists(PROD, schema, name);
  if (!exists) return { exists: false, count: 0, json: "[]" };
  // SELECT specific columns into json
  const colJsonPairs = cols
    .map((c) => `${sqlString(c)}, t.${sqlIdent(c)}`)
    .join(", ");
  const sql = `
    SELECT
      COALESCE(jsonb_agg(jsonb_build_object(${colJsonPairs})), '[]'::jsonb)::text AS data,
      count(*) AS n
    FROM ${sqlIdent(schema)}.${sqlIdent(name)} t
  `;
  const r = await query(PROD, sql);
  return {
    exists: true,
    count: Number(r?.[0]?.n ?? 0),
    json: r?.[0]?.data ?? "[]",
  };
}

async function loadTable(schema, name, cols, json) {
  const exists = await tableExists(SANDBOX, schema, name);
  if (!exists) return { skipped: true, reason: "not on sandbox" };
  const colList = cols.map(sqlIdent).join(", ");
  const sql = `
    SET session_replication_role = replica;
    DELETE FROM ${sqlIdent(schema)}.${sqlIdent(name)};
    INSERT INTO ${sqlIdent(schema)}.${sqlIdent(name)} (${colList})
    SELECT ${colList} FROM jsonb_populate_recordset(
      NULL::${sqlIdent(schema)}.${sqlIdent(name)},
      ${sqlString(json)}::jsonb
    );
    SET session_replication_role = origin;
  `;
  await query(SANDBOX, sql);
  const r = await query(
    SANDBOX,
    `SELECT count(*) AS n FROM ${sqlIdent(schema)}.${sqlIdent(name)}`,
  );
  return { loaded: Number(r?.[0]?.n ?? 0) };
}

async function main() {
  console.log(`Mirroring prod (${PROD}) → sandbox (${SANDBOX})\n`);
  const summary = [];
  for (const t of TABLES) {
    process.stdout.write(`  ${t.schema}.${t.name} … `);
    try {
      // Use the SANDBOX column list (we need cols that exist on
      // sandbox; prod columns are a superset or equal). If prod
      // has extra cols we don't know about on sandbox, they're
      // discarded — that's the right behavior.
      const sExists = await tableExists(SANDBOX, t.schema, t.name);
      if (!sExists) {
        console.log("skipped (not on sandbox)");
        summary.push({ table: `${t.schema}.${t.name}`, status: "skipped: sandbox missing" });
        continue;
      }
      const cols = await insertableColumns(SANDBOX, t.schema, t.name);
      if (cols.length === 0) {
        console.log("skipped (no insertable columns)");
        summary.push({ table: `${t.schema}.${t.name}`, status: "skipped: no cols" });
        continue;
      }
      const d = await dumpTable(t.schema, t.name, cols);
      if (!d.exists) {
        console.log("skipped (not on prod)");
        summary.push({ table: `${t.schema}.${t.name}`, status: "skipped: prod missing" });
        continue;
      }
      const r = await loadTable(t.schema, t.name, cols, d.json);
      if (r.skipped) {
        console.log(`skipped: ${r.reason}`);
        summary.push({ table: `${t.schema}.${t.name}`, status: `skipped: ${r.reason}` });
      } else {
        console.log(`${d.count} → ${r.loaded}`);
        summary.push({
          table: `${t.schema}.${t.name}`,
          status: d.count === r.loaded ? "ok" : `MISMATCH ${d.count}→${r.loaded}`,
        });
      }
    } catch (e) {
      console.log(`FAILED: ${e.message.slice(0, 220)}`);
      summary.push({
        table: `${t.schema}.${t.name}`,
        status: `error: ${e.message.slice(0, 100)}`,
      });
    }
  }
  console.log("\nSummary:");
  for (const s of summary) console.log(`  ${s.table.padEnd(28)} ${s.status}`);
  const failed = summary.filter((s) => /error|MISMATCH/.test(s.status));
  if (failed.length > 0) {
    console.log(`\n${failed.length} issue(s).`);
    process.exit(1);
  }
}

main().catch((e) => {
  console.error("FATAL:", e);
  process.exit(1);
});

File: scripts/mirror-prod-to-sandbox.py
#!/usr/bin/env python3
"""One-time snapshot mirror of production Supabase data into sandbox.

Reads PRODUCTION (strictly SELECT-only) and writes ONLY to SANDBOX. Uses
the Supabase Management API `database/query` endpoint, so it needs only a
personal access token (SUPABASE_ACCESS_TOKEN) — no DB passwords.

  Dry run (default):  SUPABASE_ACCESS_TOKEN=... python scripts/mirror-prod-to-sandbox.py
  Execute:            SUPABASE_ACCESS_TOKEN=... python scripts/mirror-prod-to-sandbox.py --execute

Loads happen with session_replication_role='replica' so FK/trigger checks
are skipped (order-independent). Encrypted/secret + transient tables are
NOT copied. Production is never modified.
"""
import os
import sys
import json
import urllib.request
import urllib.error

TOKEN = os.environ.get("SUPABASE_ACCESS_TOKEN")
if not TOKEN:
    raise SystemExit("SUPABASE_ACCESS_TOKEN not set")

PROD = "bwtmtpcgilvrkhougjdo"      # rokki-production  (READ ONLY)
SANDBOX = "hqsdhwlokfwcitfitees"   # rokki-staging     (write target)
assert PROD != SANDBOX, "refusing to run with prod==sandbox"

EXECUTE = "--execute" in sys.argv
TAG = "$MIRRORDATA$"  # dollar-quote tag; assumed absent from the data

# User-facing tables to mirror. Order is irrelevant (replica mode). Each
# is DELETE'd then re-filled from prod, so sandbox matches prod exactly.
COPY = [
    "public.profiles",
    "public.spaces",
    "public.space_members",
    "public.terminals",
    "public.terminal_members",
    "public.folders",
    "public.tasks",
    "public.task_assignees",
    "public.subtasks",
    "public.task_watchers",
    "public.calendar_connections",
    "public.calendar_events",
    "public.message_threads",
    "public.messages",
    "public.activity",
]
# Deliberately NOT copied: domain_events, rate_limit_hits,
# session_revocations, impersonation_events (transient / audit logs);
# access_tokens, push_subscriptions (per-env secrets; empty anyway);
# platform_config, modules_catalog, feature_flags (env config — leave
# sandbox's own).


def api(ref, sql):
    req = urllib.request.Request(
        f"https://api.supabase.com/v1/projects/{ref}/database/query",
        data=json.dumps({"query": sql}).encode(),
        headers={
            "Authorization": f"Bearer {TOKEN}",
            "Content-Type": "application/json",
            "User-Agent": "curl/8.5.0",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=180) as r:
            return json.loads(r.read().decode())
    except urllib.error.HTTPError as e:
        raise RuntimeError(f"HTTP {e.code}: {e.read().decode()[:400]}")


def read_prod(sql):
    s = sql.lstrip().lower()
    assert s.startswith("select"), "prod is read-only"
    return api(PROD, sql)


def write_sandbox(sql):
    # Hard guard: writes can only ever target SANDBOX.
    return api(SANDBOX, sql)


def count(ref, tbl):
    return api(ref, f"SELECT count(*)::int AS c FROM {tbl}")[0]["c"]


def table_columns(tbl):
    """(insertable cols, generated cols). Generated (e.g. search_vector
    tsvector) can't take an explicit value — Postgres recomputes them."""
    schema, name = tbl.split(".")
    rows = read_prod(
        "SELECT column_name, is_generated FROM information_schema.columns "
        f"WHERE table_schema='{schema}' AND table_name='{name}' "
        "ORDER BY ordinal_position"
    )
    insertable = [r["column_name"] for r in rows if r["is_generated"] != "ALWAYS"]
    generated = [r["column_name"] for r in rows if r["is_generated"] == "ALWAYS"]
    return insertable, generated


def fetch_json(ref, sql):
    return read_prod(sql)[0]["d"] if ref == PROD else api(ref, sql)[0]["d"]


print(f"Mirror  prod({PROD})  ->  sandbox({SANDBOX})   EXECUTE={EXECUTE}\n")

# ---- 1) make sure every prod user exists in sandbox (add missing) -------
prod_users = read_prod(
    "SELECT coalesce(json_agg(u), '[]'::json) AS d FROM auth.users u"
)[0]["d"]
sandbox_ids = {r["id"] for r in api(SANDBOX, "SELECT id FROM auth.users")}
missing = [u for u in prod_users if u["id"] not in sandbox_ids]
print(
    f"auth.users  prod={len(prod_users)}  sandbox={len(sandbox_ids)}  "
    f"missing={[u['email'] for u in missing]}"
)
if missing and EXECUTE:
    # Minimal insert — just enough for FK integrity. The display name
    # comes from the copied public.profiles row; the user won't log in
    # to sandbox, so password/identity columns aren't needed.
    vals = ",".join(
        "('00000000-0000-0000-0000-000000000000'::uuid, "
        f"'{u['id']}'::uuid, 'authenticated', 'authenticated', "
        f"'{u['email']}', now(), now())"
        for u in missing
    )
    try:
        write_sandbox(
            "SET session_replication_role='replica'; "
            "INSERT INTO auth.users "
            "(instance_id, id, aud, role, email, created_at, updated_at) "
            f"VALUES {vals} ON CONFLICT (id) DO NOTHING; "
            "SET session_replication_role='origin';"
        )
        print("  + added missing user(s)")
    except Exception as e:  # noqa: BLE001
        print(f"  ! could not add missing user(s) (continuing): {e}")

# ---- 2) copy each table -------------------------------------------------
print()
failures = []
for tbl in COPY:
    try:
        insertable, generated = table_columns(tbl)
        rows = read_prod(
            f"SELECT coalesce(json_agg(t), '[]'::json) AS d FROM {tbl} t"
        )[0]["d"]
        # Drop generated columns from the payload so json_populate_recordset
        # never tries to set them.
        for r in rows:
            for g in generated:
                r.pop(g, None)
        before = count(SANDBOX, tbl)
        line = f"{tbl:<28} prod={len(rows):<4} sandbox_before={before:<4}"
        if not EXECUTE:
            print(line + "  (dry-run)")
            continue
        j = json.dumps(rows)
        assert TAG not in j, f"{TAG} present in {tbl} data"
        cols = ", ".join(f'"{c}"' for c in insertable)
        write_sandbox(
            "SET session_replication_role='replica'; "
            f"DELETE FROM {tbl}; "
            f"INSERT INTO {tbl} ({cols}) "
            f"SELECT {cols} FROM json_populate_recordset(null::{tbl}, {TAG}{j}{TAG}::json); "
            "SET session_replication_role='origin';"
        )
        after = count(SANDBOX, tbl)
        ok = "OK" if after == len(rows) else "MISMATCH!"
        print(line + f" -> sandbox_after={after:<4} {ok}")
    except urllib.error.HTTPError as e:
        msg = e.read().decode()[:200]
        print(f"{tbl:<28} FAILED: HTTP {e.code} {msg}")
        failures.append((tbl, msg))
    except Exception as e:  # noqa: BLE001
        print(f"{tbl:<28} FAILED: {e}")
        failures.append((tbl, str(e)))

print()
if failures:
    print(f"{len(failures)} table(s) failed:")
    for t, m in failures:
        print(f"  - {t}: {m}")
    sys.exit(1)
print("done" + ("" if EXECUTE else "  (dry run — no writes performed)"))
